File: app/src/test/java/com/example/myapplication/core/nutrition/NutritionTargetCalculatorTest.kt
package com.example.myapplication.core.nutrition

import com.example.myapplication.core.profile.ActivityLevel
import com.example.myapplication.core.profile.GoalPace
import com.example.myapplication.core.profile.MetabolicSex
import com.example.myapplication.core.profile.PersonalProfile
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDate
import kotlin.math.abs

class NutritionTargetCalculatorTest {
    private val calculator = NutritionTargetCalculator()

    @Test
    fun `male profile calculates deterministic gradual loss target`() {
        val target = calculator.calculate(profile = maleProfile(), ageYears = 31).requireTarget()

        assertEquals(1724, target.basalCalories)
        assertEquals(2672, target.maintenanceCalories)
        assertEquals(2405, target.calories)
        // Weight-based macros: 78kg * 2.0g = 156g protein, 78kg * 0.8g = 62.4g (62g) fat
        assertEquals(156, target.proteinGrams)
        assertEquals(62, target.fatGrams)
        // Carbs remaining: (2404.63 - 156*4 - 62.4*9)/4 = (2404.63 - 624 - 561.6)/4 = 304.76 (305g carbs)
        assertEquals(305, target.carbsGrams)
        val macroCalories = target.proteinGrams * 4 + target.carbsGrams * 4 + target.fatGrams * 9
        assertTrue(abs(target.calories - macroCalories) <= 10)
        assertEquals(1723.75, target.audit.rawBasalCalories, 0.001)
        assertEquals(2671.8125, target.audit.rawMaintenanceCalories, 0.001)
    }

    @Test
    fun `female profile uses the female Mifflin St Jeor constant`() {
        val profile = maleProfile().copy(
            metabolicSex = MetabolicSex.FEMALE,
            heightCm = 165.0,
            currentWeightKg = 65.0,
            targetWeightKg = 55.0,
            activityLevel = ActivityLevel.MODERATE,
        )

        val target = calculator.calculate(profile, ageYears = 30).requireTarget()

        assertEquals(1370, target.basalCalories)
        assertEquals(2124, target.maintenanceCalories)
    }

    @Test
    fun `weight gain and maintenance use the expected target direction`() {
        val gain = calculator.calculate(
            maleProfile().copy(targetWeightKg = 82.0),
            ageYears = 31,
        ).requireTarget()
        val maintenance = calculator.calculate(
            maleProfile().copy(targetWeightKg = 78.0),
            ageYears = 31,
        ).requireTarget()

        // Weight gain uses +5% for MILD: 2672 * 1.05 = 2805 kcal
        assertEquals(2805, gain.calories)
        assertEquals(2672, maintenance.calories)
    }

    @Test
    fun `standard pace uses a larger initial adjustment than gradual pace`() {
        val gradual = calculator.calculate(maleProfile(), ageYears = 31).requireTarget()
        val standard = calculator.calculate(
            maleProfile().copy(goalPace = GoalPace.STANDARD),
            ageYears = 31,
        ).requireTarget()

        assertEquals(2405, gradual.calories)
        assertEquals(2271, standard.calories)
    }

    @Test
    fun `automatic weekly adjustment is capped by five percent or 150 calories`() {
        assertEquals(120, calculator.capAutomaticCalorieDelta(currentCalories = 2400, requestedDelta = 400))
        assertEquals(-100, calculator.capAutomaticCalorieDelta(currentCalories = 2000, requestedDelta = -400))
        assertEquals(150, calculator.capAutomaticCalorieDelta(currentCalories = 4000, requestedDelta = 400))
        assertEquals(40, calculator.capAutomaticCalorieDelta(currentCalories = 2000, requestedDelta = 40))
    }

    @Test
    fun `invalid profile or age requires professional review`() {
        val invalidProfile = maleProfile().copy(personalizationConsent = false)

        assertTrue(calculator.calculate(invalidProfile, ageYears = 31) is CalculationResult.NeedsProfessionalReview)
        assertTrue(calculator.calculate(maleProfile(), ageYears = 17) is CalculationResult.NeedsProfessionalReview)
    }

    @Test
    fun `timeline faster than point nine kilograms per week requires review`() {
        val today = LocalDate.of(2026, 7, 2)
        val result = calculator.calculate(
            profile = maleProfile().copy(targetWeightKg = 68.0),
            ageYears = 31,
            timeline = TargetTimeline(
                todayEpochDay = today.toEpochDay(),
                targetDateEpochDay = today.plusWeeks(4).toEpochDay(),
            ),
        )

        assertTrue(result is CalculationResult.NeedsProfessionalReview)
    }

    @Test
    fun `reasonable timeline remains calculable`() {
        val today = LocalDate.of(2026, 7, 2)
        val result = calculator.calculate(
            profile = maleProfile(),
            ageYears = 31,
            timeline = TargetTimeline(
                todayEpochDay = today.toEpochDay(),
                targetDateEpochDay = today.plusWeeks(16).toEpochDay(),
            ),
        )

        assertTrue(result is CalculationResult.Target)
    }

    private fun CalculationResult.requireTarget(): NutritionTarget {
        assertTrue(this is CalculationResult.Target)
        return (this as CalculationResult.Target).value
    }

    private fun maleProfile() = PersonalProfile(
        birthDateEpochDay = LocalDate.of(1995, 6, 15).toEpochDay(),
        metabolicSex = MetabolicSex.MALE,
        heightCm = 175.0,
        currentWeightKg = 78.0,
        targetWeightKg = 72.0,
        activityLevel = ActivityLevel.MODERATE,
        goalPace = GoalPace.MILD,
        personalizationConsent = true,
        cloudAiConsent = false,
    )
}

File: flutter/test/core/nutrition/nutrition_target_calculator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_app/core/model/nutrition_models.dart';
import 'package:gym_app/core/model/profile_models.dart';
import 'package:gym_app/core/nutrition/nutrition_target_calculator.dart';

void main() {
  final calculator = NutritionTargetCalculator();

  int day(int year, int month, int day) {
    final date = DateTime.utc(year, month, day);
    return date.millisecondsSinceEpoch ~/ (24 * 60 * 60 * 1000);
  }

  PersonalProfile maleProfile() {
    return PersonalProfile(
      birthDateEpochDay: day(1995, 6, 15),
      metabolicSex: MetabolicSex.male,
      heightCm: 175.0,
      currentWeightKg: 78.0,
      targetWeightKg: 72.0,
      activityLevel: ActivityLevel.moderate,
      goalPace: GoalPace.mild,
      personalizationConsent: true,
      cloudAiConsent: false,
    );
  }

  NutritionTarget requireTarget(CalculationResult result) {
    return result.when(
      target: (value) => value,
      needsProfessionalReview: (reason) =>
          fail("Expected target result, but got needsProfessionalReview: $reason"),
    );
  }

  bool isNeedsProfessionalReview(CalculationResult result) {
    return result.when(
      target: (_) => false,
      needsProfessionalReview: (_) => true,
    );
  }

  bool isTarget(CalculationResult result) {
    return result.when(
      target: (_) => true,
      needsProfessionalReview: (_) => false,
    );
  }

  test('male profile calculates deterministic gradual loss target', () {
    final target = requireTarget(calculator.calculate(profile: maleProfile(), ageYears: 31));

    expect(target.basalCalories, 1724);
    expect(target.maintenanceCalories, 2672);
    expect(target.calories, 2405);
    // Weight-based macros: 78kg * 2.0g = 156g protein, 78kg * 0.8g = 62.4g (62g) fat
    expect(target.proteinGrams, 156);
    expect(target.fatGrams, 62);
    // Carbs remaining: (2404.63 - 156*4 - 62.4*9)/4 = (2404.63 - 624 - 561.6)/4 = 304.76 (305g carbs)
    expect(target.carbsGrams, 305);
    final macroCalories = target.proteinGrams * 4 + target.carbsGrams * 4 + target.fatGrams * 9;
    expect((target.calories - macroCalories).abs() <= 10, isTrue);
    expect(target.audit.rawBasalCalories, closeTo(1723.75, 0.001));
    expect(target.audit.rawMaintenanceCalories, closeTo(2671.8125, 0.001));
  });

  test('female profile uses the female Mifflin St Jeor constant', () {
    final profile = maleProfile().copyWith(
      metabolicSex: MetabolicSex.female,
      heightCm: 165.0,
      currentWeightKg: 65.0,
      targetWeightKg: 55.0,
      activityLevel: ActivityLevel.moderate,
    );

    final target = requireTarget(calculator.calculate(profile: profile, ageYears: 30));

    expect(target.basalCalories, 1370);
    expect(target.maintenanceCalories, 2124);
  });

  test('weight gain and maintenance use the expected target direction', () {
    final gain = requireTarget(
      calculator.calculate(
        profile: maleProfile().copyWith(targetWeightKg: 82.0),
        ageYears: 31,
      ),
    );
    final maintenance = requireTarget(
      calculator.calculate(
        profile: maleProfile().copyWith(targetWeightKg: 78.0),
        ageYears: 31,
      ),
    );

    // Weight gain uses +5% for MILD: 2672 * 1.05 = 2805 kcal
    expect(gain.calories, 2805);
    expect(maintenance.calories, 2672);
  });

  test('standard pace uses a larger initial adjustment than gradual pace', () {
    final gradual = requireTarget(calculator.calculate(profile: maleProfile(), ageYears: 31));
    final standard = requireTarget(
      calculator.calculate(
        profile: maleProfile().copyWith(goalPace: GoalPace.standard),
        ageYears: 31,
      ),
    );

    expect(gradual.calories, 2405);
    expect(standard.calories, 2271);
  });

  test('automatic weekly adjustment is capped by five percent or 150 calories', () {
    expect(calculator.capAutomaticCalorieDelta(2400, 400), 120);
    expect(calculator.capAutomaticCalorieDelta(2000, -400), -100);
    expect(calculator.capAutomaticCalorieDelta(4000, 400), 150);
    expect(calculator.capAutomaticCalorieDelta(2000, 40), 40);
  });

  test('invalid profile or age requires professional review', () {
    final invalidProfile = maleProfile().copyWith(personalizationConsent: false);

    expect(isNeedsProfessionalReview(calculator.calculate(profile: invalidProfile, ageYears: 31)), isTrue);
    expect(isNeedsProfessionalReview(calculator.calculate(profile: maleProfile(), ageYears: 17)), isTrue);
  });

  test('timeline faster than point nine kilograms per week requires review', () {
    final todayEpoch = day(2026, 7, 2);
    final targetDateEpoch = todayEpoch + 4 * 7; // plus 4 weeks
    final result = calculator.calculate(
      profile: maleProfile().copyWith(targetWeightKg: 68.0),
      ageYears: 31,
      timeline: TargetTimeline(
        todayEpochDay: todayEpoch,
        targetDateEpochDay: targetDateEpoch,
      ),
    );

    expect(isNeedsProfessionalReview(result), isTrue);
  });

  test('reasonable timeline remains calculable', () {
    final todayEpoch = day(2026, 7, 2);
    final targetDateEpoch = todayEpoch + 16 * 7; // plus 16 weeks
    final result = calculator.calculate(
      profile: maleProfile(),
      ageYears: 31,
      timeline: TargetTimeline(
        todayEpochDay: todayEpoch,
        targetDateEpochDay: targetDateEpoch,
      ),
    );

    expect(isTarget(result), isTrue);
  });
}
